pkg/metricsadapter/provider/resourcemetrics.go

```diff
@@ -0,0 +1,463 @@
+	labelSelector, err := labels.Parse(selector)
+	if err != nil {
+		klog.Errorf("Failed to parse label selector: %v", err)
+		return nil, err
+	}
+
+	clusters, err := r.clusterLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("Failed to list clusters: %v", err)
+		return nil, err
+	}
+
+	// step 1. Find out the target clusters with lister cache
+	var targetClusters []string
+	nodesGVR := corev1.SchemeGroupVersion.WithResource("nodes")
+	for _, cluster := range clusters {
+		sci := r.informerManager.GetSingleClusterManager(cluster.Name)
+		if sci == nil {
+			klog.Errorf("Failed to get cluster(%s) manager", cluster.Name)
+			continue
+		}
+		nodes, err := sci.Lister(nodesGVR).List(labelSelector)
+		if err != nil {
+			klog.Errorf("Failed to list nodes in cluster(%s): %v", cluster.Name, err)
+			continue
+		}
+		if len(nodes) != 0 {
+			targetClusters = append(targetClusters, cluster.Name)
+		}
+	}
+
+	// step 2. Query metrics from the filtered target clusters
+	var nodeMetrics []metrics.NodeMetrics
+	for _, clusterName := range targetClusters {
+		sci := r.informerManager.GetSingleClusterManager(clusterName)
+		if sci == nil {
+			klog.Errorf("Failed to get cluster(%s) manager", clusterName)
+			continue
+		}
+
+		metricsV1betaGVR := metricsv1beta1.SchemeGroupVersion.WithResource("nodes")
+		metricsList, err := sci.GetClient().Resource(metricsV1betaGVR).List(context.Background(), metav1.ListOptions{
+			LabelSelector: selector,
+		})
+		if err != nil {
+			klog.Errorf("Failed to list node metrics in cluster(%s): %v", clusterName, err)
+			continue
+		}
+		internalMetrics, err := metricsConvertV1beta1NodeToInternalNode(metricsList.Items...)
+		if err != nil {
+			return nil, err
+		}
+
+		nodeMetrics = append(nodeMetrics, internalMetrics...)
+	}
+
+	return nodeMetrics, nil
+}
+
+// GetPodMetrics queries metrics by the internal constructed pod
+func (r *ResourceMetricsProvider) GetPodMetrics(pods ...*metav1.PartialObjectMetadata) ([]metrics.PodMetrics, error) {
+	var podMetrics []metrics.PodMetrics
+
+	// In the previous step, we construct pods list with only one element.
+	if len(pods) != 1 {
+		return podMetrics, nil
+	}
+
+	// In the previous step, if query with label selector, the name will be set to empty
+	if pods[0].Name == "" {
+		namespace := pods[0].Labels[namespaceLabelInternal]
+		selectorStr := pods[0].Labels[labelSelectorLabelInternal]
+		return r.queryPodMetricsBySelector(selectorStr, namespace)
+	}
+
+	return r.queryPodMetricsByName(pods[0].Name, pods[0].Namespace)
+}
+
+// GetNodeMetrics queries metrics by the internal constructed node
+func (r *ResourceMetricsProvider) GetNodeMetrics(nodes ...*corev1.Node) ([]metrics.NodeMetrics, error) {
+	var nodeMetrics []metrics.NodeMetrics
+
+	// In the previous step, we construct node list with only one element.
+	if len(nodes) != 1 {
+		return nodeMetrics, nil
+	}
+
+	// In the previous step, if query with label selector, the name will be set to empty
```

Comment:
I have some doubts. There is no `node` object in the Karmada control panel, so why are there some metrics related to nodes returned here.

```
kubectl --kubeconfig karmada.kubeconfig get --raw /apis/metrics.k8s.io/v1beta1/nodes 
# Some information will be returned.
```

```
kubectl  --kubeconfig karmada.kubeconfig get nodes  
# No resources found 
```

Reply:
The HPA controller will exclusively utilize the `/apis/metrics.k8s.io/v1beta1/pods` API. Therefore, it is acceptable if the other API is not available. In my opinion, users should use `karmadactl top` to inquire about metrics.

However, if we do provide the metrics API, we cannot prevent users from using it. If we choose not to provide it, there will be a need for an explanation.
cc @RainbowMango for checking

Reply:
I prefer to keep the node metrics as we are unsure if users will utilize them.